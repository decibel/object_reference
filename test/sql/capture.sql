\set ECHO none

\i test/load.sql

/*
 * SEE ALSO dump/load_all.sql and dump/verify.sql!
 */

CREATE SCHEMA test_support;
SET search_path = test_support, tap, public;

\i test/helpers/object_table.sql

CREATE SCHEMA object_identity_temp_test_schema; -- THIS NEEDS TO FAIL IF THE SCHEMA EXISTS!
SET search_path=object_identity_temp_test_schema,test_support, tap, public;

CREATE TABLE object_group_ids(n int, object_group_id int);
CREATE TEMP VIEW og_o AS
  SELECT n, ogo.*
    FROM _object_reference.object_group__object ogo
      JOIN object_group_ids ogi USING(object_group_id)
;

SELECT plan( (
  0
  
  + (SELECT count(*) FROM test_prereq)
  + 1 -- capture__stop when not capturing

  + 2 -- create groups

  + 2 -- start capture

  + c -- create

  + 3 -- Manually register an object; make certain it does not show up in group

  + cna + 2 -- verify #1

  + 3 -- Group 2

  + 1 -- Invalid capture group

  + 3 -- Stop group 2
  + 3 -- Stop group 1

  -- drop
  + 2 -- Drop groups
    + cna * 2 -- Drop objects
    + 1 -- Verify object_group_ids still has correct count
    + 1 -- verify object table is now empty
)::int )
  FROM (SELECT count(*) c, count(CASE WHEN create_command NOT LIKE 'ALTER%' THEN 1 END) AS cna
    FROM test_object) c
;

SELECT lives_ok(
      command
      , 'prereq: ' || command
    )
  FROM test_prereq
;

SELECT throws_ok(
  $$SELECT object_reference.capture__stop('')$$
  , 'P0002'
  , 'object group "" does not exist'
  , 'Verify capture__stop() without capture__start errors.'
);

-- Create groups
SELECT lives_ok(
  format(
    $$INSERT INTO object_group_ids VALUES(
        %1$s
        , object_reference.object_group__create('TEMP object capture test group %1$s')
      )
    $$
    , n
  )
  , format('Create "TEMP object capture test group %s"', n)
) FROM generate_series(1,2) n(n);

-- Start capture
SELECT is(
  object_reference.capture__start('TEMP object capture test group 1')
  , 1
  , 'Start capture for group 1'
);
SELECT is_empty(
  $$SELECT * FROM og_o$$
  , 'No objects exist for either test group'
);

-- Create
SELECT c.* FROM test_object o, test__create(o) c ORDER BY o.seq ASC;

-- Manually register an object; make certain it does not show up in group
-- NOTE! Other tests depend on this working!
SELECT lives_ok(
  $$CREATE TEMP TABLE ogi__object_id AS
      SELECT object_reference.object__getsert('table', 'object_group_ids', NULL) AS object_id
  $$
  , 'Register object_group_ids table'
);
SELECT is(
  (SELECT count(*) FROM ogi__object_id)
  , 1::bigint
  , 'Verify ogi__object_id has exactly one row.'
);
SELECT is_empty(
  $$SELECT row_to_json(og_o.*)
      FROM _object_reference.object_group__object og_o
        JOIN ogi__object_id USING(object_id)
  $$
  , 'Verify object_id for object_group_ids table is NOT in any groups.'
);

-- Verify #1
SELECT is(
  (SELECT count(*) FROM og_o WHERE n=1)
  , (SELECT count(*) FROM test_object
      WHERE create_command NOT LIKE 'ALTER%'
  )
  , 'Correct # of objects captured to group.'
);
SELECT c.*
  FROM test_object o, test__register(o) c
  WHERE create_command NOT LIKE 'ALTER%'
  ORDER BY o.seq
;
SELECT bag_eq(
  $$SELECT object_id FROM og_o WHERE n = 1$$
  , $$SELECT object_id FROM obj_ref$$
  , 'Verify captured object IDs match'
);
/*
SELECT * FROM og_o;
SELECT * FROM _object_reference.object;-- WHERE object_id IN(6,9);
*/

-- Group 2
SELECT is(
  object_reference.capture__start('TEMP object capture test group 2')
  , 2
  , 'Start capture for group 2'
);
CREATE TABLE "Capture 2"();
CREATE VIEW "Capture 2 View" AS SELECT 1;
CREATE TEMP TABLE cap2 AS SELECT * FROM (VALUES
  ('table'::cat_tools.object_type, '"Capture 2"')
  , ('view', '"Capture 2 View"')
) v(object_type, object_name)
;

SELECT bag_eq(
  $$SELECT object_id FROM og_o WHERE n = 2$$
  , $$SELECT object_reference.object__getsert(object_type, object_name, NULL) FROM cap2$$
  , 'Verify captured objects for level 2'
);
SELECT bag_eq(
  $$SELECT object_id FROM og_o WHERE n = 1$$
  , $$SELECT object_id FROM obj_ref$$
  , 'Verify group 1 objects have not changed.'
);

-- Invalid capture group
SELECT throws_ok(
  $$SELECT object_reference.capture__stop('TEMP object capture test group DOES NOT EXIST')$$
  , 'P0002' -- Should probably change this...
  , 'object group "TEMP object capture test group DOES NOT EXIST" does not exist'
  , 'capture__stop() with invalid group errors'
);

-- Stop group 2
SELECT lives_ok(
  $$SELECT object_reference.capture__stop('TEMP object capture test group 2')$$
  , 'Stop capture group 2'
);

CREATE TABLE "Capture 1"();
SELECT bag_eq(
  $$SELECT object_id FROM og_o WHERE n = 1$$
  , $$SELECT object_id FROM obj_ref
    UNION SELECT object_reference.object__getsert('table', '"Capture 1"', NULL )
  $$
  , 'Verify table "Capture 1" added to capture group 1'
);
SELECT bag_eq(
  $$SELECT object_id FROM og_o WHERE n = 2$$
  , $$SELECT object_reference.object__getsert(object_type, object_name, NULL) FROM cap2$$
  , 'Verify captured objects for level 2 are unchanged'
);

-- Stop group 1
SELECT lives_ok(
  $$SELECT object_reference.capture__stop('TEMP object capture test group 1')$$
  , 'Stop capture group 1'
);
CREATE TABLE "Capture 3"();
SELECT bag_eq(
  $$SELECT object_id FROM og_o WHERE n = 1$$
  , $$SELECT object_id FROM obj_ref
    UNION SELECT object_reference.object__getsert('table', '"Capture 1"', NULL )
  $$
  , 'Verify capture group 1 is still correct'
);
SELECT bag_eq(
  $$SELECT object_id FROM og_o WHERE n = 2$$
  , $$SELECT object_reference.object__getsert(object_type, object_name, NULL) FROM cap2$$
  , 'Verify capture group 2 is still correct'
);

--SET client_min_messages = DEBUG;
SELECT lives_ok(
  format( $$SELECT object_reference.object_group__remove(%s, true)$$, n )
  , 'Remove object group ' || n
) FROM object_group_ids
;
SELECT c.*
  FROM test_object o, test__drop(o) c
  WHERE create_command NOT LIKE 'ALTER%'
  ORDER BY o.seq DESC -- NEEDS to be DESC!
;

SELECT is(
  (SELECT count(*)::int FROM object_group_ids)
  , 2
  , 'object_group_ids still has correct record count'
);
  
SELECT is_empty(
  $$SELECT * FROM og_o$$
  , 'No object references remain'
);

\i test/pgxntool/finish.sql

-- vi: expandtab sw=2 ts=2
