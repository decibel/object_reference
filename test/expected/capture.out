\set ECHO none
1..69
ok 1 - prereq: CREATE DOMAIN "test domain" int
ok 2 - prereq: CREATE FUNCTION tg_null() RETURNS trigger LANGUAGE plpgsql AS $body$BEGIN RETURN NEW; END$body$
ok 3 - prereq: CREATE TYPE "test type"
ok 4 - prereq: CREATE FUNCTION "test type in"(cstring) RETURNS "test type" LANGUAGE 'internal' IMMUTABLE AS 'int2in'
ok 5 - prereq: CREATE FUNCTION "test type out"("test type") RETURNS cstring LANGUAGE 'internal' IMMUTABLE AS 'int2in'
ok 6 - Verify capture__stop() without capture__start errors.
ok 7 - Create "TEMP object capture test group 1"
ok 8 - Create "TEMP object capture test group 2"
ok 9 - Start capture for group 1
ok 10 - No objects exist for either test group
ok 11 - CREATE table "test table" ("test column" int)
ok 12 - CREATE index "test table test index" ON "test table"("test column")
ok 13 - CREATE sequence "test sequence"
ok 14 - CREATE view "test view" AS SELECT * FROM "test table"
ok 15 - CREATE materialized view "test materialized view" AS SELECT * FROM "test table"
ok 16 - ALTER TABLE "test table" ADD COLUMN "second test column" int
ok 17 - ALTER DOMAIN "test domain" ADD CONSTRAINT "test domain constraint" CHECK(true)
ok 18 - ALTER TABLE "test table" ADD CONSTRAINT "test table constraint" CHECK(true)
ok 19 - CREATE FUNCTION "test function"("test column" int DEFAULT 0) RETURNS int LANGUAGE sql AS 'SELECT $1'
ok 20 - CREATE TYPE "test type" (INPUT = "test type in", OUTPUT = "test type out")
ok 21 - CREATE CAST ("test type" AS int4) WITH INOUT
ok 22 - ALTER TABLE "test table" ALTER "test column" SET DEFAULT 0
ok 23 - CREATE TRIGGER "test trigger" AFTER INSERT ON "test table" FOR EACH ROW EXECUTE PROCEDURE tg_null()
ok 24 - Register object_group_ids table
ok 25 - Verify ogi__object_id has exactly one row.
ok 26 - Verify object_id for object_group_ids table is NOT in any groups.
ok 27 - Correct # of objects captured to group.
ok 28 - Register: INSERT INTO obj_ref VALUES(1, object_reference.object__getsert('table', '"test table"', NULL))
ok 29 - Register: INSERT INTO obj_ref VALUES(2, object_reference.object__getsert('index', '"test table test index"', NULL))
ok 30 - Register: INSERT INTO obj_ref VALUES(3, object_reference.object__getsert('sequence', '"test sequence"', NULL))
ok 31 - Register: INSERT INTO obj_ref VALUES(4, object_reference.object__getsert('view', '"test view"', NULL))
ok 32 - Register: INSERT INTO obj_ref VALUES(5, object_reference.object__getsert('materialized view', '"test materialized view"', NULL))
ok 33 - Register: INSERT INTO obj_ref VALUES(9, object_reference.object__getsert('function', '"test function"', '"test column" int DEFAULT 0'))
ok 34 - Register: INSERT INTO obj_ref VALUES(10, object_reference.object__getsert('type', '"test type"', NULL))
ok 35 - Register: INSERT INTO obj_ref VALUES(11, object_reference.object__getsert('cast', '"test type"', 'integer'))
ok 36 - Register: INSERT INTO obj_ref VALUES(13, object_reference.object__getsert('trigger', '"test table"', 'test trigger'))
ok 37 - Verify captured object IDs match
ok 38 - Start capture for group 2
ok 39 - Verify captured objects for level 2
ok 40 - Verify group 1 objects have not changed.
ok 41 - capture__stop() with invalid group errors
ok 42 - Stop capture group 2
ok 43 - Verify table "Capture 1" added to capture group 1
ok 44 - Verify captured objects for level 2 are unchanged
ok 45 - Stop capture group 1
ok 46 - Verify capture group 1 is still correct
ok 47 - Verify capture group 2 is still correct
ok 48 - Remove object group 1
ok 49 - Remove object group 2
ok 50 - Drop should fail while reference exists
ok 51 - DROP TRIGGER "test trigger" ON "test table"
ok 52 - Drop should fail while reference exists
ok 53 - DROP CAST ("test type" AS int4)
ok 54 - Drop should fail while reference exists
ok 55 - DROP TYPE "test type" CASCADE; -- Need to cascade due to functions
ok 56 - Drop should fail while reference exists
ok 57 - DROP FUNCTION "test function"(int)
ok 58 - Drop should fail while reference exists
ok 59 - DROP materialized view "test materialized view"
ok 60 - Drop should fail while reference exists
ok 61 - DROP view "test view"
ok 62 - Drop should fail while reference exists
ok 63 - DROP sequence "test sequence"
ok 64 - Drop should fail while reference exists
ok 65 - DROP index "test table test index"
ok 66 - Drop should fail while reference exists
ok 67 - DROP table "test table"
ok 68 - object_group_ids still has correct record count
ok 69 - No object references remain
# TRANSACTION INTENTIONALLY LEFT OPEN!
